backend/src/main/java/com/redhat/cloud/notifications/routers/internal/kessel/KesselAssetsMigrationService.java

```diff
@@ -0,0 +1,144 @@
+import io.grpc.stub.StreamObserver;
+import io.quarkus.logging.Log;
+import jakarta.annotation.security.RolesAllowed;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import jakarta.ws.rs.BadRequestException;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Path;
+import org.project_kessel.api.relations.v1beta1.CreateTuplesRequest;
+import org.project_kessel.api.relations.v1beta1.CreateTuplesResponse;
+import org.project_kessel.api.relations.v1beta1.ObjectReference;
+import org.project_kessel.api.relations.v1beta1.ObjectType;
+import org.project_kessel.api.relations.v1beta1.Relationship;
+import org.project_kessel.api.relations.v1beta1.SubjectReference;
+import org.project_kessel.relations.client.RelationTuplesClient;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.redhat.cloud.notifications.Constants.API_INTERNAL;
+
+@ApplicationScoped
+@RolesAllowed(ConsoleIdentityProvider.RBAC_INTERNAL_ADMIN)
+@Path(API_INTERNAL)
+public class KesselAssetsMigrationService {
+    /**
+     * Defines the relation between the subject and the resource when passing
+     * the tuple to Kessel. <a href="https://github.com/RedHatInsights/rbac-config/blob/a806fb03c95959391eceb0b42c7eefd8ae2350ae/configs/prod/schemas/schema.zed#L3">
+     * Reference</a>
+     */
+    public static final String RELATION = "t_workspace";
+
+    @Inject
+    BackendConfig backendConfig;
+
+    @Inject
+    EndpointRepository endpointRepository;
+
+    @Inject
+    RelationTuplesClient relationTuplesClient;
+
+    @Inject
+    StatusRepository statusRepository;
+
+    @Inject
+    WorkspaceUtils workspaceUtils;
+
+    @Path("/kessel/migrate-assets")
+    @POST
+    public void migrateAssets() {
+        // We need the application to be in maintenance mode to avoid any
+        // updates to the database, which could cause the Notifications
+        // database and the Kessel inventory to be out of sync.
+        if (Status.MAINTENANCE != this.statusRepository.getCurrentStatus().getStatus()) {
+            throw new BadRequestException("Unable to execute migration because the application is not in maintenance mode");
+        }
+
+        int fetchedEndpointsSize = 0;
+        int offset = 0;
+        do {
+            final List<Endpoint> fetchedEndpoints = this.endpointRepository.getEndpointsWithLimitAndOffset(this.backendConfig.getKesselMigrationBatchSize(), offset);
+
+            // If for some reason we have fetched full pages from the database
+            // all the time, the last one might be empty, so there is no need
+            // to attempt calling Kessel.
+            if (fetchedEndpoints.isEmpty()) {
+                break;
+            }
+
+            final CreateTuplesRequest request = this.createTuplesRequest(fetchedEndpoints);
+            final int finalOffset = offset;
+            this.relationTuplesClient.createTuples(request, new StreamObserver<>() {
```

Comment:
Is this service replayable? Will it be ok if we send the same endpoint multiple times?

Reply:
I have been told that in principle yes, but that they will tell us for sure soon.

Reply:
It is not replayable, so we will only be able to run the migration once. We will have to make sure we disable "creates" and "deletes" for existing integrations.